### PR TITLE
feat(robustness): add liveness check and readiness check

### PR DESCRIPTION
**Pull request by JasonHe-WQ**
## What type of PR is this?

/kind feature

## What this PR does / why we need it:

For better robustness

## Which issue(s) this PR fixes:

Issue #462 

The connectivity of etcd and APIServer seems to be a [build-in feature](https://github.com/kubernetes/apiserver/blob/dcd23e3027703a1a63efcd851b4571c4fb445f63/pkg/server/options/etcd.go#L327)

### CODE REVIEW

**Comment by JasonHe-WQ** [comment]
I have read the CLA Document and I hereby sign the CLA

File: pkg/core/health/check.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+// used to check if the application is ready to serve traffic.
+func NewReadyzHandler(serv *server.KarporServer, sync *syncer.ResourceSyncer, sg storage.Storage) http.HandlerFunc {
+	conf := HandlerConfig{
+		Verbose: true,
+		// checkList is a list of healthcheck to run.
+		Checks: []Check{
+			NewPingCheck(),
+			NewServerCheck(serv),
+			NewSyncerCheck(sync),
+			NewStorageCheck(sg),
```
Comment (elliotxx) [inline pkg/core/health/check.go:59]: Here can be simplified:
```go
			NewPingCheck(),
			NewETCDCheck(serv),
			NewStorageCheck(sg),
```
If you are not sure how to probe ETCD, you can continue to simplify it as:
```go
			NewPingCheck(),
			NewStorageCheck(sg),
```

File: pkg/infra/persistence/elasticsearch/client.go [hunk 2]
```diff
@@ -417,3 +417,22 @@ func (cl *Client) termsAgg(ctx context.Context, index string, field string) (*Ag
...
 }
+
+// PingElasticSearchClient would ping ElasticSearch client
+func (cl *Client) PingElasticSearchClient(ctx context.Context) error {
+	req := esapi.PingRequest{
+		Pretty: true,
+	}
+
+	// 使用带有上下文的请求
+	res, err := req.Do(ctx, cl.client)
```
Comment (elliotxx) [inline pkg/infra/persistence/elasticsearch/client.go:428]: You can use `es.client.Info()` and `es.client.Cluster.Health()` instead directly, example:
```go
func checkElasticsearchLiveness(es *elasticsearch.Client) error {
    // 检查连接
    if _, err := es.Info() // 使用 Info 方法获取服务信息
    {
        return fmt.Errorf("failed to connect to Elasticsearch")
    }

    // 检查集群健康状态
    resp, err := es.Cluster.Health(nil)
    if err != nil {
        return fmt.Errorf("failed to get cluster health: %v", err)
    }
    defer resp.Body.Close()

    if status := resp.IsStatus("green", "yellow"); !status {
        return fmt.Errorf("Elasticsearch cluster health is not green or yellow")
    }

    return nil
}
```

File: pkg/core/route/route.go [hunk 3]
```diff
@@ -108,6 +109,8 @@ func NewCoreRoute(
 	// Endpoint to list all available endpoints in the router.
 	router.Get("/server-configs", expvar.Handler().ServeHTTP)
 
+	// TODO: Add health check endpoints
+	health.Register(router, nil, nil)
```
Comment (elliotxx) [inline pkg/core/route/route.go:113]: Can TODO be removed?

File: pkg/core/health/server_check.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	}
+}
+
+func (s *ServerCheck) Pass(ctx context.Context) bool {
+	return s.HealthCheck(ctx) == nil
+}
+
+func (s *ServerCheck) Name() string {
+	return "ServerCheck"
+}
```
Comment (elliotxx) [inline pkg/core/health/server_check.go:40]: Need to remove this file. Server Check is actually defined in NewReadyzHandler.
```go
Checks: []Check{
    NewPingCheck(),
    NewStorageCheck(sg),
},
```

File: go.mod [hunk 5]
```diff
@@ -150,12 +150,12 @@ require (
...
 	go.uber.org/atomic v1.7.0 // indirect
 	go.uber.org/zap v1.24.0 // indirect
-	golang.org/x/arch v0.0.0-20201008161808-52c3e6f60cff // indirect
+	golang.org/x/arch v0.3.0 // indirect
 	golang.org/x/crypto v0.16.0 // indirect
 	golang.org/x/mod v0.14.0 // indirect
 	golang.org/x/net v0.19.0 // indirect
 	golang.org/x/oauth2 v0.0.0-20220223155221-ee480838109b // indirect
-	golang.org/x/sync v0.5.0 // indirect
+	golang.org/x/sync v0.5.0
```
Comment (elliotxx) [inline go.mod:158]: Need to move to `direct` import. Then execute `go mod tidy` again.

File: pkg/infra/search/storage/types.go [hunk 6]
```diff
@@ -60,13 +61,15 @@ type ResourceGroupRuleStorage interface {
...
 	ListResourceGroupsBy(ctx context.Context, ruleName string) (*ResourceGroupResult, error)
+	CheckHealth
 }
 
 // SearchStorage interface defines the basic operations for search storage.
 type SearchStorage interface {
 	Search(ctx context.Context, queryString, patternType string, pagination *Pagination) (*SearchResult, error)
 	SearchByTerms(ctx context.Context, keysAndValues map[string]any, pagination *Pagination) (*SearchResult, error)
 	AggregateByTerms(ctx context.Context, keys []string) (*AggregateResults, error)
+	CheckHealth
```
Comment (elliotxx) [inline pkg/infra/search/storage/types.go:72]: Just add `CheckHealth` to storage

File: pkg/core/health/check.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	"github.com/KusionStack/karpor/pkg/infra/search/storage"
+	"github.com/go-chi/chi/v5"
+	"net/http"
+)
+
+type StorageImpl struct {
+	storage.ResourceStorage
+	storage.SearchStorage
+	storage.ResourceGroupRuleStorage
+}
```
Comment (elliotxx) [inline pkg/core/health/check.go:29]: The `Storage` structure has implemented the `Storage` interface. No `StorageImpl` is required.
`Storage` structure: pkg/infra/search/storage/elasticsearch/client.go

File: pkg/infra/search/storage/elasticsearch/factory.go [hunk 8]
```diff
@@ -83,6 +83,27 @@ func (s *ResourceGroupRuleStorageGetter) GetResourceGroupRuleStorage() (storage.
 	return esClient, nil
 }
 
+// GeneralStorageGetter creates a new instance of the
+// GeneralStorageGetter with the given Elasticsearch addresses, user name, and
+// password.
+type GeneralStorageGetter struct {
```
Comment (elliotxx) [inline pkg/infra/search/storage/elasticsearch/factory.go:89]: The comment  here are a bit strange, can you optimize it?

**Review by elliotxx — APPROVED** [review]
lgtm

**Pull request merged**